src/Data/Object/Query.elm

```diff
@@ -78,18 +79,65 @@ removeComponent id ({ components } as query) =
     }
 
 
-updateComponentItem : Component.Item -> Query -> Query
-updateComponentItem newItem query =
+updateComponentItem : Component.Id -> (Item -> Item) -> List Item -> List Item
+updateComponentItem componentId fn =
+    List.map
+        (\item ->
+            if item.id == componentId then
+                fn item
+
+            else
+                item
+        )
+
+
+updateComponentItemQuantity : Component.Id -> Component.Quantity -> Query -> Query
+updateComponentItemQuantity id quantity query =
     { query
         | components =
             query.components
-                |> List.map
-                    (\item ->
-                        if item.id == newItem.id then
-                            newItem
+                |> updateComponentItem id (\item -> { item | quantity = quantity })
+    }
+
 
-                        else
-                            item
+updateElementAmount : Component -> Int -> Component.Amount -> Query -> Query
+updateElementAmount component index amount query =
+    let
+        updateElements =
+            List.indexedMap
+                (\i element ->
+                    if i == index then
+                        { element | amount = amount }
+
+                    else
+                        element
+                )
+    in
+    { query
+        | components =
+            query.components
+                |> updateComponentItem component.id
+                    (\item ->
+                        { item
+                            | custom =
+                                case item.custom of
+                                    Just custom ->
+                                        let
+                                            updatedCustom =
+                                                { custom | elements = updateElements custom.elements }
+                                        in
+                                        if Component.isCustomized component updatedCustom then
+                                            Just updatedCustom
+
+                                        else
+                                            Nothing
+
+                                    Nothing ->
+                                        Just
+                                            { elements = updateElements component.elements
+                                            , name = Nothing
+                                            }
+                        }
                     )
```

Comment:
Maybe replacing it with a local function like the `updateElements` one could enhance the readability?
And it looks like the `case` could be replaced by a `Maybe.withDefault`?

Reply:
I swear I tested many refactors but this version is the more legible one I could find. I'm all ears for your take on this, but otherwise I'm just gonna land this as is.

Reply:
slightly improved legibility in 0dfb7d12b08c4e320a622398cb239227fa4f9195